src/main/java/tech/stackable/hbase/opa/OpaAclChecker.java

```diff
@@ -0,0 +1,190 @@
+  private final Optional<Cache<String, Boolean>> aclCache;
+
+  public static class CacheConfig {
+    public final boolean enabled;
+    public final int ttlSeconds;
+    public final long maxSize;
+
+    public CacheConfig(boolean enabled, int ttlSeconds, long maxSize) {
+      this.enabled = enabled;
+      this.ttlSeconds = ttlSeconds;
+      this.maxSize = maxSize;
+    }
+  }
+
+  public OpaAclChecker(
+      boolean authorizationEnabled, String opaPolicyUrl, boolean dryRun, CacheConfig cc) {
+    this.authorizationEnabled = authorizationEnabled;
+    this.dryRun = dryRun;
+    this.aclCache =
+        cc.enabled
+            ? Optional.of(
+                Caffeine.newBuilder()
+                    .expireAfterWrite(cc.ttlSeconds, TimeUnit.SECONDS)
+                    .maximumSize(cc.maxSize)
+                    .build())
+            : Optional.empty();
+
+    this.json =
+        new ObjectMapper()
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+            .configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false)
+            .setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE)
+            .setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.PUBLIC_ONLY)
+            .setVisibility(PropertyAccessor.GETTER, JsonAutoDetect.Visibility.PUBLIC_ONLY);
+
+    if (authorizationEnabled) {
+      if (opaPolicyUrl == null) {
+        throw new OpaException.UriMissing(OpenPolicyAgentAccessController.OPA_POLICY_URL_PROP);
+      }
+
+      try {
+        this.opaUri = URI.create(opaPolicyUrl);
+      } catch (Exception e) {
+        throw new OpaException.UriInvalid(opaUri, e);
+      }
+    }
+  }
+
+  public void checkPermissionInfo(User user, TableName table, Permission.Action action)
+      throws AccessControlException {
+    OpaAllowQuery query =
+        new OpaAllowQuery(new OpaAllowQuery.OpaAllowQueryInput(user.getUGI(), table, action));
+    this.checkPermissionInfo(query);
+  }
+
+  public void checkPermissionInfo(User user, String namespace, Permission.Action action)
+      throws AccessControlException {
+    OpaAllowQuery query =
+        new OpaAllowQuery(new OpaAllowQuery.OpaAllowQueryInput(user.getUGI(), namespace, action));
+    this.checkPermissionInfo(query);
+  }
+
+  public void checkPermissionInfo(OpaAllowQuery query) throws AccessControlException {
+    if (!this.authorizationEnabled) {
+      return;
+    }
+
+    String body;
+    try {
+      body = json.writeValueAsString(query);
+    } catch (JsonProcessingException e) {
+      throw new OpaException.SerializeFailed(e);
+    }
+
+    String prettyPrinted;
+    try {
+      prettyPrinted = json.writerWithDefaultPrettyPrinter().writeValueAsString(query);
+    } catch (JsonProcessingException e) {
+      LOG.error(
+          "Could not pretty print the following request body (printing raw version instead): {}",
+          body);
+      throw new OpaException.SerializeFailed(e);
+    }
```

Comment:
It seems redundant to me to pretty print this to me.
I'm pretty (sic!) sure that you can configure Jackson to directly pretty print instead of doing a second pass. But this could be called hundreds of thousands of times per second so it needs to be as fast as possible.